block/gcs/adapter.go

```diff
@@ -0,0 +1,438 @@
+	httpClient            *http.Client
+	ctx                   context.Context
+	uploadIDTranslator    block.UploadIDTranslator
+	streamingChunkSize    int
+	streamingChunkTimeout time.Duration
+}
+
+func WithHTTPClient(c *http.Client) func(a *Adapter) {
+	return func(a *Adapter) {
+		a.httpClient = c
+	}
+}
+
+func WithStreamingChunkSize(sz int) func(a *Adapter) {
+	return func(a *Adapter) {
+		a.streamingChunkSize = sz
+	}
+}
+
+func WithStreamingChunkTimeout(d time.Duration) func(a *Adapter) {
+	return func(a *Adapter) {
+		a.streamingChunkTimeout = d
+	}
+}
+
+func WithContext(ctx context.Context) func(a *Adapter) {
+	return func(a *Adapter) {
+		a.ctx = ctx
+	}
+}
+
+func WithTranslator(t block.UploadIDTranslator) func(a *Adapter) {
+	return func(a *Adapter) {
+		a.uploadIDTranslator = t
+	}
+}
+
+func NewAdapter(s3 s3iface.S3API, opts ...func(a *Adapter)) block.Adapter {
+	a := &Adapter{
+		s3:                    s3,
+		httpClient:            http.DefaultClient,
+		ctx:                   context.Background(),
+		uploadIDTranslator:    &block.NoOpTranslator{},
+		streamingChunkSize:    DefaultStreamingChunkSize,
+		streamingChunkTimeout: DefaultStreamingChunkTimeout,
+	}
+	for _, opt := range opts {
+		opt(a)
+	}
+	return a
+}
+
+func (s *Adapter) WithContext(ctx context.Context) block.Adapter {
+	return &Adapter{
+		s3:                    s.s3,
+		httpClient:            s.httpClient,
+		ctx:                   ctx,
+		uploadIDTranslator:    s.uploadIDTranslator,
+		streamingChunkSize:    s.streamingChunkSize,
+		streamingChunkTimeout: s.streamingChunkTimeout,
+	}
+}
+
+func (s *Adapter) log() logging.Logger {
+	return logging.FromContext(s.ctx)
+}
+
+// work around, because put failed with trying to create symlinks
+func (s *Adapter) PutWithoutStream(obj block.ObjectPointer, sizeBytes int64, reader io.Reader, opts block.PutOpts) error {
+	qualifiedKey, err := resolveNamespace(obj)
+	if err != nil {
+		return err
+	}
+	putObject := s3.PutObjectInput{
+		Body:         aws.ReadSeekCloser(reader),
+		Bucket:       aws.String(qualifiedKey.StorageNamespace),
+		Key:          aws.String(qualifiedKey.Key),
+		StorageClass: opts.StorageClass,
+	}
+	_, err = s.s3.PutObject(&putObject)
+	return err
+}
+
+func (s *Adapter) Put(obj block.ObjectPointer, sizeBytes int64, reader io.Reader, opts block.PutOpts) error {
+	var err error
+	defer reportMetrics("Put", time.Now(), &sizeBytes, &err)
+
+	qualifiedKey, err := resolveNamespace(obj)
+	if err != nil {
+		return err
+	}
+	putObject := s3.PutObjectInput{
+		Bucket:       aws.String(qualifiedKey.StorageNamespace),
+		Key:          aws.String(qualifiedKey.Key),
+		StorageClass: opts.StorageClass,
+	}
+	sdkRequest, _ := s.s3.PutObjectRequest(&putObject)
+	_, err = s.streamRequestData(sdkRequest, sizeBytes, reader)
+	return err
+}
+
+func (s *Adapter) UploadPart(obj block.ObjectPointer, sizeBytes int64, reader io.Reader, uploadID string, partNumber int64) (string, error) {
+	var err error
```

Comment:
This doesn't work, right? Maybe throw Unimplemented?

Reply:
Same for other multipart-related stuff

Reply:
I'll keep it as ref implmentation